core/arch/arm/kernel/tee_ta_manager.c

```diff
@@ -236,373 +242,224 @@ static struct tee_ta_ctx *tee_ta_context_find(const TEE_UUID *uuid)
-	size += sizeof(ta_head_t) +
-		sec_head->nbr_func * sizeof(ta_func_head_t) +
-		(sec_head->rel_dyn_got_size & TA_HEAD_GOT_MASK);
-
-	ctx->mm = tee_mm_alloc(&tee_mm_sec_ddr, size);
-	if (!ctx->mm)
-		return TEE_ERROR_OUT_OF_MEMORY;
-
-	/*
-	 * Copy TA into reserved memory space (DDR).
+	 * Map physical memory into TA virtual memory
 	 */
 
 	res = tee_mmu_init(ctx);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto out;
 
 	res = tee_mmu_map(ctx, &param);
-	if (res != TEE_SUCCESS) {
-		EMSG("call tee_mmu_map_uta() failed %X", res);
-		return res;
-	}
+	if (res != TEE_SUCCESS)
+		goto out;
 
 	tee_mmu_set_ctx(ctx);
 
-	dst = (void *)tee_mmu_get_load_addr(ctx);
-	if (!tee_vbuf_is_non_sec(nmem_ta, size)) {
-		EMSG("User TA isn't in non-secure memory");
-		return TEE_ERROR_SECURITY;
+	res = elf_load_body(elf_state, tee_mmu_get_load_addr(ctx));
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	digest = malloc(shdr->hash_size);
+	if (!digest) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
 	}
-	memcpy(dst, nmem_ta, size);
 
-	cache_maintenance_l1(DCACHE_AREA_CLEAN, dst, size);
-	cache_maintenance_l1(ICACHE_AREA_INVALIDATE, dst, size);
+	res = crypto_ops.hash.final(hash_ctx, hash_algo, digest,
+				    shdr->hash_size);
+	if (res != TEE_SUCCESS)
+		goto out;
 
-	ctx->load_addr = tee_mmu_get_load_addr(ctx);
+	if (memcmp(digest, SHDR_GET_HASH(shdr), shdr->hash_size) != 0)
+		res = TEE_ERROR_SECURITY;
 
-	return TEE_SUCCESS;
+	cache_maintenance_l1(DCACHE_AREA_CLEAN,
+			     (void *)tee_mmu_get_load_addr(ctx), vasize);
+	cache_maintenance_l1(ICACHE_AREA_INVALIDATE,
+			     (void *)tee_mmu_get_load_addr(ctx), vasize);
+out:
+	elf_load_final(elf_state);
+	free(digest);
+	free(hash_ctx);
+	return res;
 }
 
-static void tee_ta_load_init_user_ta(struct tee_ta_ctx *ctx, size_t heap_size)
-{
-	/* Init rel.dyn, GOT, ZI and heap */
-	tee_ta_init_reldyn(ctx);
-	tee_ta_init_got(ctx);
-	tee_ta_init_heap(ctx, heap_size);
-	tee_ta_init_zi(ctx);
-}
+
 
 /*-----------------------------------------------------------------------------
  * Loads TA header and hashes.
  * Verifies the TA signature.
  * Returns session ptr and TEE_Result.
  *---------------------------------------------------------------------------*/
-static TEE_Result tee_ta_load(const kta_signed_header_t *signed_ta,
-			      struct tee_ta_ctx **ta_ctx)
+static TEE_Result tee_ta_load(const TEE_UUID *uuid,
+			const struct shdr *signed_ta,
+			struct tee_ta_ctx **ta_ctx)
 {
-	/* ta & ta_session is assumed to be != NULL from previous checks */
 	TEE_Result res;
-	size_t heap_size = 0;	/* gcc warning */
+	/* man_flags: mandatory flags */
+	uint32_t man_flags = TA_FLAG_USER_MODE | TA_FLAG_EXEC_DDR;
+	/* opt_flags: optional flags */
+	uint32_t opt_flags = man_flags | TA_FLAG_SINGLE_INSTANCE |
+	    TA_FLAG_MULTI_SESSION | TA_FLAG_UNSAFE_NW_PARAMS;
 	struct tee_ta_ctx *ctx = NULL;
-	const uint8_t *nmem_ta;
-	kta_signed_header_t *sec_signed_ta = NULL;
-	ta_head_t *sec_head = NULL;
+	struct shdr *sec_signed_ta = NULL;
```

Comment:
I would rename `sec_signed_ta` in `sec_signed_header_ta` as it will contain only the "sign/hash" part of the TA.